round-based/src/mpc/mod.rs

```diff
@@ -0,0 +1,329 @@
+    async fn complete<R>(
+        &mut self,
+        round: Self::Round<R>,
+    ) -> Result<R::Output, Self::CompleteRoundErr<R::Error>>
+    where
+        R: RoundInfo,
+        Self::Msg: RoundMsg<R::Msg>;
+
+    /// Sends a message
+    ///
+    /// This method awaits until the message is sent, which might be not the best method to use if you
+    /// need to send many messages at once. If it's the case, prefer using [`.send_many()`](Self::send_many).
+    async fn send(&mut self, msg: Outgoing<Self::Msg>) -> Result<(), Self::SendErr>;
+
+    /// Sends a p2p message to another party
+    ///
+    /// Note: when you send many messages at once (it's most likely the case when you send a p2p message), this method
+    /// is not efficient, prefer using [`.send_many()`](Self::send_many).
+    async fn send_p2p(
+        &mut self,
+        recipient: PartyIndex,
+        msg: Self::Msg,
+    ) -> Result<(), Self::SendErr> {
+        self.send(Outgoing::p2p(recipient, msg)).await
+    }
+
+    /// Sends a message that will be received by all parties
+    ///
+    /// Message will be broadcasted, but not reliably. If you need a reliable broadcast, use
+    /// [`MpcExecution::reliably_broadcast`] method.
+    async fn send_to_all(&mut self, msg: Self::Msg) -> Result<(), Self::SendErr> {
+        self.send(Outgoing::all_parties(msg)).await
+    }
+
+    /// Reliably broadcasts a message
+    ///
+    /// Message will be received by all participants of the protocol. Moreover, when recipient receives a
+    /// message, it will be assured (cryptographically or through other trust assumptions) that all honest
+    /// participants of the protocol received the same message.
+    ///
+    /// It's a responsibility of a message delivery layer to provide the reliable broadcast mechanism. If
+    /// it's not supported, this method returns an error. Note that not every MPC protocol requires the
+    /// reliable broadcast, so it's totally normal to have a message delivery implementation that does
+    /// not support it.
+    async fn reliably_broadcast(&mut self, msg: Self::Msg) -> Result<(), Self::SendErr> {
+        self.send(Outgoing::reliable_broadcast(msg)).await
+    }
+
+    /// Creates a buffer of outgoing messages so they can be sent all at once
+    ///
+    /// When you have many messages that you want to send at once, using [`.send()`](Self::send) is not efficient,
+    /// as it will accumulate message delivery cost. Use this method to optimize sending many messages at once.
+    fn send_many(self) -> Self::SendMany;
+
+    /// Yields execution
+    async fn yield_now(&self);
+}
+
+/// Buffer, optimized for sending many messages at once
+pub trait SendMany {
+    /// MPC executor, returned after successful [`.flush()`](Self::flush)
+    type Exec: MpcExecution<Msg = Self::Msg, SendErr = Self::SendErr>;
+    /// Protocol message
+    type Msg;
+    /// Error indicating that sending a message has failed
+    type SendErr;
```

Comment:
I'm not sure, but can we simply use `Self::Exec::Msg` and `Self::Exec::SendErr` in this trait? Or the compiler can't resolve that?

Reply:
It can, but you have to write `<Self::Exec as MpcExecution>::Msg`, etc. It's common practice to include associated type like this, e.g. see [`IntoIterator`](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html) and [`IntoInterator::Item`](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html#associatedtype.Item) which is actually `IntoIterator::IntoIter::Item`